socketio_middleware/src/rooms.rs

```diff
@@ -33,35 +35,114 @@ impl ChannelPair {
 }
 
 pub fn join_channel_to_room(room_id: &str, channel_pair: ChannelPair) {
+    //TODO: why use remove, it is low performance.
     let mut connected_sockets = match ROOMS.remove(room_id) {
         Some(val) => val,
         None => Vec::new(),
     };
 
-    connected_sockets.push(channel_pair);
+    //check if socketid exist
+    let mut exist: bool = false;
+    for socket in &connected_sockets {
+        if socket.sid() == channel_pair.sid() {
+            debug!("ROOMS: socketid {} doesn't join into room {}, this socketid already exist in the room.", channel_pair.sid(), room_id);
+            exist = true;
+            break;
+        }
+    }
 
+    if false == exist {
+        debug!("ROOMS: socketid {} joined into room {}.", channel_pair.sid(), room_id);
+        connected_sockets.push(channel_pair);
+    }
     ROOMS.insert(room_id.to_string(), connected_sockets);
+
+    //print all sockets in the room
+    match ROOMS.get(room_id) {
+        Some(sockets) => {
+            for socket in &*sockets {
+                debug!("ROOMS: room {} containted socketid {}, sockets number = {}.", room_id, socket.sid(), sockets.len());
+            }
+        }
+        None => {
+            debug!("ROOMS: no socketid in room {}.", room_id);
+        }
+    }
 }
 
-pub fn remove_socket_from_room(room_id: &str, _sid: &str) {
+/*
+pub fn join_channel_to_room(room_id: &str, channel_pair: ChannelPair) {
+    match ROOMS.get_mut(room_id) {
+        Some(mut connected_sockets) => {
+            //check if socketid exist
+            for socket in &*connected_sockets {
+                debug!("In room {}, socket_id = {}", room_id, socket.sid());
+                if socket.sid() == channel_pair.sid() {
+                    debug!("Join channel to room {}, socket_id({}) is exist.", room_id, socket.sid());
+                    //return;
+                }
+            }
+
+            debug!("Socket_id {} insert a exist room {}", channel_pair.sid(), room_id);
+            connected_sockets.push(channel_pair);
+            ROOMS.insert(room_id.to_string(), connected_sockets.to_vec());
+        }
+        None => {
+            debug!("Socket_id {} insert a new room {}", channel_pair.sid(), room_id);
+
+            let mut connected_sockets = Vec::new();
+            connected_sockets.push(channel_pair);
+            ROOMS.insert(room_id.to_string(), connected_sockets);
+        }
+    };
+}
+*/
+
+pub fn remove_socket_from_room(room_id: &str, sid: &str) {
     let mut connected_sockets = match ROOMS.remove(room_id) {
         Some(val) => val,
         None => Vec::new(),
     };
 
-    for i in 0..connected_sockets.len() - 1 {
-        let i = connected_sockets.len() - 1 - i;
-        let socket = connected_sockets.get(i).unwrap();
-
-        if socket.sid == room_id {
+    //check if socketid exist
+    let mut i = 0;
+    for socket in &connected_sockets {
```

Comment:
let's refactor this bit to be closer to the original,

```
for let i = 0..connected_sockets.len()  {
  let socket = connected_sockets[i];
  etc...
}
```

Reply:
Yes, I use the new loop.